File: backend/constants/schemas/trip.py
from schema import Schema, Or
from constants.schemas.main import metadata_schema

stop_schema = Schema({
    "Code": str,
    "Order": int,
    "Time": str,
    "sortingTime": str,
    "IsMajor": bool
})

trip_schema = Schema({
    "Number": str,
    "Display": str,
    "Line": str,
    "Direction": str,
    "LineVariant": str,
    "Type": str,
    "Stops": {
        "Stop": [stop_schema]
    },
    "destinationStopCode": str,
    "departFromCode": str,
    "departFromAlternativeCode": str,
    "departFromTimingPoint": str,
    "tripPatternId": Or(int, str)
})

transfer_schema = Schema({
    "Code": str,
    "Order": int,
    "Time": str
})

transfer_link_schema = Schema({
    "FromTrip": str,
    "FromStopCode": str,
    "ToTrip": str,
    "ToStopCode": str,
    "TransferDuration": str
})

service_schema = Schema({
    "Colour": str,
    "Type": str,
    "Direction": str,
    "Code": str,
    "StartTime": str,
    "EndTime": str,
    "Duration": str,
    "Accessible": "",
    "Trips": {
        "Trip": [trip_schema]
    },
    "Transfers": {
        "Transfer": [transfer_schema]
    },
    "TransferLinks": {
        "Link": [transfer_link_schema]
    },
    "StartSortTime": str,
    "EndSortTime": str,
    "tripHash": str,
    "transferCount": int
})

journey_schema = Schema({
    "Date": str,
    "Time": str,
    "To": str,
    "From": str,
    "Services": Or([service_schema], None)
})

trips_schema = Schema({
    "Metadata": metadata_schema,
    "SchJourneys": [journey_schema]
})

File: frontend/src/types/trip.ts
import { z } from "zod";
import { StopSchema } from "./stops";

// ---- Stop Schema ----
const stopSchema = z.object({
    Code: z.string(),
    Order: z.number(),
    Time: z.string(),
    sortingTime: z.string(),
    IsMajor: z.boolean(),
});

// ---- Trip Schema ----
const TripSchema = z.object({
    display: z.string(),
    line: z.string(),
    lineVariant: z.string(),
    type: z.string(),
    Stops: z.object({
        Stop: z.array(stopSchema),
    }),
    destinationStopCode: z.string(),
    departFromCode: z.string(),
    departFromAlternativeCode: z.string(),
    departFromTimingPoint: z.string(),
    tripPatternId: z.union([z.string(), z.number()]),
});

// ---- Transfer Schema ----
const transferSchema = z.object({
  Code: z.string(),
  Order: z.number(),
  Time: z.string(),
});

// ---- Transfer Link Schema ----
const transferLinkSchema = z.object({
    FromTrip: z.string(),
    FromStopCode: z.string(),
    ToTrip: z.string(),
    ToStopCode: z.string(),
    TransferDuration: z.string(),
});

const ServiceSchema = z.object({
    type: z.enum(['R', 'B', 'RB']),
    startTime: z.string(),
    endTime: z.string(),
    duration: z.string(),
    trips: z.array(TripSchema),
    transfers: z.array(transferSchema),
    link: z.array(transferLinkSchema),
    transferCount: z.number(),
});

export const JourneySchema = z.object({
    From: StopSchema,
    Journey: z.object({
        To: StopSchema,
        Services: z.union([z.array(ServiceSchema), z.null()])
    })
});

export type Journey = z.infer<typeof JourneySchema>;
